File: web/src/types/libraryReport.ts
import type { DuplicateCandidateGroup } from "./duplicate";

export type LibraryReportTrack = {
  id: number;
  title: string;
  artist: string | null;
  album: string | null;
  duration_seconds: number | null;
  content_type: string;
  status: string;
  updated_at: string;
  last_played_at: string | null;
  playback_count: number;
};

export type LibraryReportTrackIssue = {
  track: LibraryReportTrack;
  reasons: string[];
};

export type LibraryOrganizationReport = {
  generated_at: string;
  untagged_ready_tracks: LibraryReportTrack[];
  missing_metadata_tracks: LibraryReportTrackIssue[];
  processing_tracks: LibraryReportTrackIssue[];
  duplicate_groups: DuplicateCandidateGroup[];
  never_played_ready_tracks: LibraryReportTrack[];
  rarely_played_ready_tracks: LibraryReportTrack[];
  stale_cooldown_tracks: LibraryReportTrackIssue[];
};

export type MediaKind = "original" | "playback" | "cover" | "temporary_video";

export type MediaReferenceIssue = {
  track_id: number;
  media_kind: MediaKind;
  path: string | null;
};

export type OrphanMediaFile = {
  media_kind: MediaKind;
  path: string;
};

export type StorageConsistencyReport = {
  generated_at: string;
  referenced_file_count: number;
  scanned_file_count: number;
  missing_references: MediaReferenceIssue[];
  unsafe_references: MediaReferenceIssue[];
  orphan_files: OrphanMediaFile[];
  scan_errors: string[];
};

File: backend/app/schemas/library_report.py
from datetime import datetime

from pydantic import BaseModel

from app.schemas.duplicate import DuplicateCandidateGroupResponse


class LibraryReportTrack(BaseModel):
    id: int
    title: str
    artist: str | None
    album: str | None
    duration_seconds: int | None
    content_type: str
    status: str
    updated_at: datetime
    last_played_at: datetime | None = None
    playback_count: int = 0


class LibraryReportTrackIssue(BaseModel):
    track: LibraryReportTrack
    reasons: list[str]


class LibraryOrganizationReport(BaseModel):
    generated_at: datetime
    untagged_ready_tracks: list[LibraryReportTrack]
    missing_metadata_tracks: list[LibraryReportTrackIssue]
    processing_tracks: list[LibraryReportTrackIssue]
    duplicate_groups: list[DuplicateCandidateGroupResponse]
    never_played_ready_tracks: list[LibraryReportTrack]
    rarely_played_ready_tracks: list[LibraryReportTrack]
    stale_cooldown_tracks: list[LibraryReportTrackIssue]
